core/chains/evm/gas/block_history_estimator.go

```diff
@@ -310,76 +337,59 @@ func (b *BlockHistoryEstimator) BumpLegacyGas(_ context.Context, originalGasPric
 	return bumpedGasPrice, gasLimit, err
 }
 
-// checkConnectivity detects if the transaction is not being included due to
-// some kind of mempool propagation or connectivity issue rather than
-// insufficiently high pricing and returns error if so
-func (b *BlockHistoryEstimator) checkConnectivity(attempts []EvmPriorAttempt) error {
+// haltBumping prevents transactions from excessively bumping if an existing attempt's price is above a configurable percentile
+// This check is required in case the transaction is not being included due to some kind of mempool propagation
+// or connectivity issue rather than insufficiently high pricing
+func (b *BlockHistoryEstimator) haltBumping(attempts []EvmPriorAttempt) error {
 	percentile := int(b.bhConfig.CheckInclusionPercentile())
-	// how many blocks since broadcast?
-	latestBlockNum := b.getCurrentBlockNum()
-	if latestBlockNum == nil {
-		b.logger.Warn("Latest block is unknown; skipping inclusion check")
-		// can't determine anything if we don't have/know latest block num yet
-		return nil
+	// Get latest CheckInclusionPercentile prices to use in the checks below
+	var maxGasPrice *assets.Wei
+	var maxTipCap *assets.Wei
+	ok := b.IfStarted(func() {
+		maxGasPrice = b.getMaxPercentileGasPrice()
+		maxTipCap = b.getMaxPercentileTipCap()
+	})
+	if !ok {
+		return errors.New("BlockHistoryEstimator is not started; do not have max gas to allow bumping")
 	}
-	expectInclusionWithinBlocks := int(b.bhConfig.CheckInclusionBlocks())
-	blockHistory := b.getBlocks()
-	if len(blockHistory) < expectInclusionWithinBlocks {
-		b.logger.Warnf("Block history in memory with length %d is insufficient to determine whether transaction should have been included within the past %d blocks", len(blockHistory), b.bhConfig.CheckInclusionBlocks())
-		return nil
+	if !b.initialFetch.Load() {
+		return errors.New("BlockHistoryEstimator has not finished the first gas estimation yet, likely because a failure on start")
+	}
+	// Return error to prevent bumping if gas price is nil or if EIP1559 is enabled and tip cap is nil
+	if maxGasPrice == nil || (b.eConfig.EIP1559DynamicFees() && maxTipCap == nil) {
+		errorMsg := fmt.Sprintf("%d percentile price is not set. This is likely because there aren't any valid transactions to estimate from. Preventing bumping until valid price is available to compare", percentile)
```

Comment:
Is this safe? This means that if head tracker is not working we can't bump txes. That seems potentially dangerous.

Reply:
Yes, I see these values can be null only if we failed to calculate things on startup, which itself means BlockHistoryEstimator cannot work.
If you don't have blocks, you can't calculate %iles, thus not set gas prices at all.

If at a later time the HeadTracker stops, we will still have these values set, just to a stale value.

I'm ok with halting bumping in such cases. There is a risk, but then we already see HeadTracker not working, no no point going crazy with bumping.
I've seen many cases where our gas_prices are way way higher than the gas fees on a block, causing the nodes to go out of funds. So not halting is actually already causing production problems.